app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala

```diff
@@ -0,0 +1,101 @@
+import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+
+class BooleanExprValidator(
+  indexLookup: Map[FormComponentId, Int],
+  wrapperType: BooleanExprWrapperType
+)(
+  idx: Int
+) {
+
+  private def extractFcIds(expr: Expr): List[FormComponentId] = expr match {
+    case Add(left, right)         => extractFcIds(left) ::: extractFcIds(right)
+    case Subtraction(left, right) => extractFcIds(left) ::: extractFcIds(right)
+    case Multiply(left, right)    => extractFcIds(left) ::: extractFcIds(right)
+    case Sum(field1)              => extractFcIds(field1)
+    case FormCtx(fcId)            => List(fcId)
+    case _                        => Nil
+  }
+
+  private def extractFcIds(expr: DateExpr): List[FormComponentId] = expr match {
+    case DateExprWithOffset(dateExpr, _, _)       => extractFcIds(dateExpr)
+    case DateFormCtxVar(FormCtx(formComponentId)) => List(formComponentId)
+    case _                                        => Nil
+  }
+
+  private def validateIdIdx(id: FormComponentId) =
+    indexLookup
+      .get(id)
+      .map { idIdx =>
+        wrapperType
+          .isForwardReference(idIdx, idx)
+          .validationResult(s"id '${id.value}' named in $wrapperType is forward reference, which is not permitted")
+      }
+      .getOrElse(Invalid(s"id '${id.value}' named in $wrapperType expression does not exist in a form"))
+
+  private def validateExprs(left: Expr, right: Expr): List[ValidationResult] =
+    List(left, right)
+      .flatMap(extractFcIds)
+      .map(validateIdIdx)
+
+  private def validateDateExprs(left: DateExpr, right: DateExpr): List[ValidationResult] =
+    List(left, right)
+      .flatMap(extractFcIds)
+      .map(validateIdIdx)
+
+  private def validateValueField(expr: Expr): List[ValidationResult] =
+    extractFcIds(expr).map(validateIdIdx)
+
+  def apply(includeIf: BooleanExpr): List[ValidationResult] = includeIf match {
+    case Equals(left, right)              => validateExprs(left, right)
+    case GreaterThan(left, right)         => validateExprs(left, right)
+    case GreaterThanOrEquals(left, right) => validateExprs(left, right)
+    case LessThan(left, right)            => validateExprs(left, right)
+    case LessThanOrEquals(left, right)    => validateExprs(left, right)
+    case Not(e)                           => apply(e)
+    case Or(left, right)                  => apply(left) ::: apply(right)
+    case And(left, right)                 => apply(left) ::: apply(right)
+    case IsFalse | IsTrue                 => List(Valid)
+    case Contains(collection, value)      => validateExprs(collection, value)
+    case In(value, _)                     => validateValueField(value)
+    case DateBefore(left, right)          => validateDateExprs(left, right)
+    case DateAfter(left, right)           => validateDateExprs(left, right)
+    case MatchRegex(value, _)             => validateValueField(value)
+  }
+}
+
+sealed trait BooleanExprWrapperType {
```

Comment:
may be call it BooleanExprForwardRefChecker for clarity ?

Reply:
Well, original idea was a type which remembers in what type was `BooleanExpr` wrapped in.

That `isForwardReference` method is convenient to have here is just an after effect. I can put that method elsewhere if  you wish?

Reply:
Ok, I get it. Lets leave it as it is then